### ISSUE TITLE
RFC: WithMetadata modifier

### ISSUE THREAD

**Issue by brysgo** [facebook/graphql#381]
WithMetadata Modifier

When wrapped a WithMetadata Modifier, a scalar field can have an optional
selection underneath it. When a selection is present, the value field holds
the scalar, when a selection is not present, the scalar is returned.

Example:
```graphql
type Name: String {
  firstName: String
  middleName: String
  lastName: String
  nickname: String
  title: String
}
```

Usage:
```graphql
...
person {
  name
}
...
```
vvvvvvvvv
```javascript
...
person: {
  name: "Princess Leia"
}
...
```
or
```graphql
...
person {
  name {
    value
    title
    firstName
    middleName
    lastName
    nickname
  }
}
...
```
vvvvvvvvv
```javascript
...
person: {
  name: {
    value: "Princess Leia",
    title: "Princess",
    firstName: "Leia",
    middleName: "Organa",
    lastName: "Solo",
    nickname: "Leia"
  }
}
...
```

See graphql/graphql-js#1089 for graphql-js example.

Would love to hear everyone's thoughts.

**Comment by brysgo** [comment]
Another way of looking at it, instead of optional metadata for a scalar, would be as an optional scalar return for an object. They both look the same and accomplish the same goal of sensible defaults, with a more detailed drill down.

**Comment by adamkl** [comment]
I think one use case where this could be valuable is the ability to include field level permissions along with the data. 

Example:
```graphql
enum ePermission {
  Create
  Read
  Update
  Delete
}
  
type ProtectedString: String {
  access: [ePermission]
}

type Person {
  firstName: ProtectedString
  lastName: ProtectedString
  nickName: ProtectedString
}
```

This would allow tailoring of the UI based on the access levels of the data requested. 

Usage:
```graphql
person {
  firstName {
    value
    access
  }
  lastName {
    value
    access
  }
  nickName {
    value
    access
  }
}
```
Data as requested by Darth Vader, parent, allowed to modify legal names of child, but not child created nickname which would be display only in the UI:
```javascript
person: {
  firstName: {
    value: "Leia",
    access: [ "Read", "Update" ]
  }
  lastName: {
    value: "Organa",
    access: [ "Read", "Update" ]
  }
  nickName: {
    value: "Princess Leia",
    access: [ "Read" ]
  }
}
```
Data as requested by Princess Leia, the child, only allowed to change nickname, other fields would be display only:
```javascript
person: {
  firstName: {
    value: "Leia",
    access: [ "Read" ]
  }
  lastName: {
    value: "Organa",
    access: [ "Read" ]
  }
  nickName: {
    value: "Princess Leia",
    access: [ "Read", "Update" ]
  }
}
```
If a different UI/screen doesn't care about the permissions, then a regular query could be used:
```graphql
person {
  firstName 
  lastName 
  nickName
}
```
Which gives the expected:
```javascript
person: {
  firstName: "Leia",
  lastName: "Organa",
  nickName: "Princess Leia"
}
```
This would allow the gradual addition of field level metadata without causing any breaking changes in existing UIs.

**Comment by stubailo** [comment]
This could be replaced with a convention, right? Like:

```graphql
# simple
{ person { name } }

# detailed
{ person { nameDetails { firstName, lastName } } }
```

Is there a big benefit to having the single name return both of the above?

**Comment by brysgo** [comment]
Yes, you could, but conceptually it isn't a different field and might crowd your schema + documentation if you had to do it for every field.

**Comment by brysgo** [comment]
Also, I think we would have to figure out something that doesn't make value a pseudo reserved word. Maybe add an underscore or two?

**Comment by rmosolgo** [comment]
To expand on @stubailo's point, recall that successful RFCs usually include an existing implementation _on top of_ the current specification. So, one way that you could demonstrate the value of this feature is by building it as client infrastructure. 

For example, if the _server_ followed a convention like the [one above](https://github.com/facebook/graphql/issues/381#issuecomment-345424250), could you create a client transform so that the _client_ has the experience suggested in this issue, but the server follows the existing spec? (That is, if `name` is selected alone, it maps to the `name` field, but if `name` has sub-selections, it's transformed into `nameDetails`, then re-transformed to `name` for the client.)

That would make this change much more compelling. And as a GraphQL implementor, I get more excited about changes which have _already_ demonstrated their advantages in the wild!

**Comment by brysgo** [comment]
That is an excellent idea! I think that is a good next step. Thanks for the suggestion @rmosolgo ! BTW, I loved using your graphql implementation when I was helping a client that was using rails move their API to graphql.

**Comment by brysgo** [comment]
Before I go down a rabbit hole, has anyone built a tool to augment only the graphql query language? Specifically, I will need to:
1. Transform the query to change:
```graphql
person {
  name {
    value
    lastName
  }
}
```
into
```graphql
person {
  name
  nameMetadata {
    lastName
  }
}
```
2. Then I will need to intercept the result and change:
```json
{
  "name": "Princess Leia",
  "nameMetadata": {
    "lastName": "Leia"
  }
}
```
into
```json
{
  "name": {
    "value": "Princess Leia",
    "lastName": "Leia"
  }
}
```
I could do network layer middleware, except with relay that would require patching the compiler. Has someone built a solution for such an experiment, or am I trailblazing?
